flash: fix #18263

StefJar opened the pull request:
stm32f4 flash code assumes that it is run after reset. In case of having a bootloader run before it may changes the flash registers. This can lead to failed read/write/erase operations.

This fix addresses the stm32f4 flash driver. It clears the OPTCR register and take care, that the driver resets the flash settings to the reset defaults.

Commit:
flash: fix #18263

stm32f4 flash code assumes that it is run after reset. In case of having a bootloader run before it may changes the flash registers. This can lead to failed read/write/erase operations.

This fix addresses the stm32f4 flash driver. It clears the OPTCR register and take care, that the driver resets the flash settings to the reset defaults.

ulfalizer commented on drivers/flash/flash_stm32.c:
```diff
@@ -329,6 +329,27 @@ static int stm32_flash_init(struct device *dev)
 	LL_AHB3_GRP1_EnableClock(LL_AHB3_GRP1_PERIPH_HSEM);
 #endif /* CONFIG_SOC_SERIES_STM32WBX */
 
+#if defined(CONFIG_SOC_SERIES_STM32F4X)
```
Could have a plain `#ifdef`.

Instead of putting the issue number into the commit title, you can add this above the `Signed-off-by:` instead (and explain what the commit does in the title instead):

    Fixes: #18263 

Then the issue will be closed automatically when this gets merged too.

nashif reviewed (changes requested):
please add a meaningful title to this PR

Commit:
drivers: flash: stm32f4: resets OPTCR register.

the current stm32f4 flash code assumes that it is run after reset.
If a programm, like a bootloader, runs before the zephyr build firmware
it can happen that the FLASH controller registers are changed. In my
case this lead to failed read/write/erase operations. This fix addresses
the stm32f4 flash driver. It clears the OPTCR register and take care,
that the driver resets the flash settings to the reset defaults.

Fixes #18263

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
drivers: flash: stm32f4: resets OPTCR register.

the current stm32f4 flash code assumes that it is run after reset. If a programm, like a bootloader, runs before the zephyr build firmware it can happen that the FLASH controller registers are changed. In my case this lead to failed read/write/erase operations. This fix addresses the stm32f4 flash driver. It clears the OPTCR register and take care, that the driver resets the flash settings to the reset defaults.

Fixes #18263

Signed-off-by: Stefan Jaritz <stefan@kokoontech.com>

Commit:
Merge branch 'master' of github.com:StefJar/zephyr
